### PR TITLE
EFA 2020-01-02 updates

### PR DESCRIPTION
**Pull request by galpress**

### CODE REVIEW

File: providers/efa/efa.c [hunk 1]
```diff
@@ -50,6 +51,7 @@ static struct verbs_context *efa_alloc_context(struct ibv_device *vdev,
 	struct ibv_device_attr_ex attr;
 	struct ibv_get_context cmd;
 	struct efa_context *ctx;
+	int qp_table_sz;
```
Comment (jgunthorpe) [inline providers/efa/efa.c]: unsigned type

File: providers/efa/efa.h [hunk 2]
```diff
@@ -24,6 +24,7 @@ struct efa_context {
 	uint32_t max_llq_size;
 	size_t cqe_size;
 	struct efa_qp **qp_table;
+	int qp_table_sz_m1;
```
Comment (jgunthorpe) [inline providers/efa/efa.h]: unsigned type, do not use signed types in bit manipulations.

File: providers/efa/verbs.c [hunk 3]
```diff
@@ -940,7 +957,10 @@ struct ibv_qp *efadv_create_qp_ex(struct ibv_context *ibvctx,
 
 	if (attr_ex->qp_type != IBV_QPT_DRIVER ||
 	    !vext_field_avail(struct efadv_qp_init_attr,
-			      driver_qp_type, inlen)) {
+			      driver_qp_type, inlen) ||
+	    efa_attr->comp_mask ||
+	    !is_reserved_cleared(efa_attr->reserved) ||
```
Comment (jgunthorpe) [inline providers/efa/verbs.c]: If a reserved is switched to an actual value then we generally expect a comp mask bit for it

File: providers/efa/efa_io_defs.h [hunk 4]
```diff
@@ -6,6 +6,12 @@
 #ifndef _EFA_IO_H_
 #define _EFA_IO_H_
 
+#define EFA_GET(ptr, type) \
+	((*(ptr) & type##_MASK) >> type##_SHIFT)
+
+#define EFA_SET(ptr, type, value) \
+	*(ptr) |= ((value) << type##_SHIFT) & type##_MASK
```
Comment (jgunthorpe) [inline providers/efa/efa_io_defs.h]: Macros should be expressions not statements, see the checkpatch warning this triggers. Add a () around it.

Comment (galpress) [reply providers/efa/efa_io_defs.h]: OK for the checkpatch ().
What do you suggest for the first part though?

File: providers/efa/efa.h (hunk 2 above)
Comment (galpress) [reply providers/efa/efa.h]: Done

File: providers/efa/efa.c (hunk 1 above)
Comment (galpress) [reply providers/efa/efa.c]: Done

File: providers/efa/verbs.c (hunk 3 above)
Comment (galpress) [reply providers/efa/verbs.c]: Agree, but a non-zero value is most likely an application bug and I prefer we catch it if possible.

File: providers/efa/efa_io_defs.h (hunk 4 above)
Comment (jgunthorpe) [reply providers/efa/efa_io_defs.h]: What first part? Put ({}) around the whole thing

Comment (galpress) [reply providers/efa/efa_io_defs.h]: Done

**Comment by juhlee-microsoft** [comment]
@galpress What is this PR for? I am catching up the changes and need more help about this PR.

**Comment by galpress** [comment]
Hi @juhlee-microsoft 
This PR introduces misc updates to the Elastic Fabric Adapter (EFA) provider, you can see individual commit messages for more details.
Do you have a specific question I can help with?

**Pull request merged**